runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/ClockSkewInterceptor.kt

```diff
@@ -0,0 +1,117 @@
+        private val CLOCK_SKEW_ERROR_CODES = listOf(
+            "RequestTimeTooSkewed",
+            "RequestExpired",
+            "RequestInTheFuture",
+        )
+
+        // Errors possibly caused by clock skew
+        private val POSSIBLE_CLOCK_SKEW_ERROR_CODES = listOf(
+            "PriorRequestNotComplete",
+            "RequestTimeout",
+            "RequestTimeoutException",
+            "InternalError",
+        )
+    }
+
+    // Clock skew to be applied to all requests
+    private val _currentSkew: AtomicRef<Duration?> = atomic(null)
+
+    /**
+     * Apply the previously-computed skew, if it's set, to the execution context before signing
+     */
+    public override suspend fun modifyBeforeSigning(context: ProtocolRequestInterceptorContext<Any, HttpRequest>): HttpRequest {
+        val logger = coroutineContext.logger<ClockSkewInterceptor>()
+
+        val skew = _currentSkew.value
+        skew?.let {
+            logger.info { "applying clock skew $skew to client" }
+            context.executionContext[HttpOperationContext.ClockSkew] = skew
+        }
+
+        context.executionContext[HttpOperationContext.ClockSkewApproximateSigningTime] = Instant.now()
+
+        return context.protocolRequest
+    }
+
+    /**
+     * After receiving a response, check if the client clock is skewed and apply a correction if necessary.
+     */
+    public override suspend fun modifyBeforeAttemptCompletion(context: ResponseInterceptorContext<Any, Any, HttpRequest, HttpResponse?>): Result<Any> {
+        val logger = coroutineContext.logger<ClockSkewInterceptor>()
+
+        val serverTime = context.protocolResponse?.header("Date")?.let {
+            Instant.fromRfc5322(it)
+        } ?: run {
+            logger.debug { "service did not return \"Date\" header, skipping skew calculation" }
+            return context.response
+        }
+
+        val clientTime = context.protocolRequest.headers["Date"]?.let {
+            Instant.fromRfc5322(it)
+        } ?: context.protocolRequest.headers["x-amz-date"]?.let {
+            Instant.fromIso8601(it)
+        } ?: context.executionContext[HttpOperationContext.ClockSkewApproximateSigningTime]
+
+        if (clientTime.isSkewed(serverTime, context.protocolResponse?.status?.description)) {
+            val skew = clientTime.until(serverTime)
+            logger.warn { "client clock ($clientTime) is skewed $skew from the server ($serverTime), applying correction" }
+            _currentSkew.getAndSet(skew)
+            context.executionContext[HttpOperationContext.ClockSkew] = skew
+
+            // Mark the exception as retryable
+            val ex = (context.response.exceptionOrNull() as? SdkBaseException)
+            ex?.sdkErrorMetadata?.attributes?.set(ErrorMetadata.Retryable, true)
+            ex?.let { return Result.failure(it) }
+        } else {
+            logger.trace { "client clock ($clientTime) is not skewed from the server ($serverTime)" }
```

Comment:
**nit**: We probably don't need this log statement to say we didn't do anything.